src/main/java/at/helpch/placeholderapi/expansion/player/util/PlayerUtil.java

```diff
@@ -0,0 +1,139 @@
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Function;
+
+public final class PlayerUtil {
+
+    private static final BlockFace[] radial = {
+        BlockFace.NORTH, BlockFace.NORTH_EAST,
+        BlockFace.EAST, BlockFace.SOUTH_EAST,
+        BlockFace.SOUTH, BlockFace.SOUTH_WEST,
+        BlockFace.WEST, BlockFace.NORTH_WEST
+    };
+
+    /**
+     * Get player's item in hand using the right method depending on the server version
+     *
+     * @param player player
+     * @return item in hand
+     * @see Player#getItemInHand() method for < 1.9
+     * @see PlayerInventory#getItemInMainHand() method for >= 1.9
+     */
+    @SuppressWarnings({"DataFlowIssue", "deprecation"})
+    public static @Nullable ItemStack getItemInMainHand(@NotNull final Player player) {
+        return VersionHelper.HAS_OFF_HAND ? player.getInventory().getItemInMainHand() : player.getItemInHand();
+    }
+
+    public static @NotNull String getLocationInfo(
+        @Nullable final Location location, @NotNull final String defaultValue,
+        @NotNull final Function<@NotNull Location, @NotNull Object> coordinateMapper
+    ) {
+        return Optional.ofNullable(location)
+            .map(coordinateMapper)
+            .map(String::valueOf)
+            .orElse(defaultValue);
+    }
+
+    public static @NotNull String getBedLocationInfo(
+        @NotNull final OfflinePlayer player, @NotNull final String defaultValue,
+        @NotNull final Function<@NotNull Location, @NotNull Object> coordinateMapper
+    ) {
+        return getLocationInfo(player.getBedSpawnLocation(), defaultValue, coordinateMapper);
+    }
+
+    public static @NotNull String getCompassLocationInfo(
+        @NotNull final Player player, @NotNull final String defaultValue,
+        @NotNull final Function<@NotNull Location, @NotNull Object> coordinateMapper
+    ) {
+        return getLocationInfo(player.getCompassTarget(), defaultValue, coordinateMapper);
+    }
+
+    public static int getEmptySlotsCount(@NotNull final Player player) {
+        final PlayerInventory inventory = player.getInventory();
+        final List<ItemStack> items = new ArrayList<>(Arrays.asList(inventory.getContents()));
+        items.add(inventory.getHelmet());
+        items.add(inventory.getChestplate());
+        items.add(inventory.getLeggings());
+        items.add(inventory.getBoots());
+
+        if (VersionHelper.HAS_OFF_HAND) {
+            items.add(inventory.getItemInOffHand());
+        }
+
+        return (int) items.stream().filter(ItemUtil::isNullOrAir).count();
+    }
+
+    public static @NotNull BlockFace getDirection(@NotNull final Player player) {
+        return radial[Math.round(player.getLocation().getYaw() / 45f) & 0x7].getOppositeFace();
+    }
+
+    public static String getXZDirection(Player player) {
+        double rotation = player.getLocation().getYaw();
+
+        if (rotation < 0.0D) {
+            rotation += 360.0D;
```

Comment:
I am unsure what values the getYaw method can return but considering that it can be lower than 0, I'd suggest we normalize the data here:
- get absolute value
- divide by 360 and get the remainder

This also prevents the multiple Math.abs calls below.

Reply:
What would `Math.abs(rotation - 90) <= 45` become then?

Reply:
> What would `Math.abs(rotation - 90) <= 45` become then?

I am a bit confused why we actually need to do that? Can't we do the Absolute, % 360, and then check the ranges and convert them too cardinal directions?

Reply:
Idk that's inherited code